pkg/worker/scm/manager.go

```diff
@@ -211,14 +213,16 @@ func CloneRepo(token string, codeSource *api.CodeSource, ref string, folder stri
 	}
 
 	var logs string
-	for i := 1; i <= maxRetry; i++ {
+	backoff := initialDelay
+	for retries := 0; retries < maxRetries; retries++ {
 		logs, err = p.Clone(token, url, reference, destPath)
 		if err == nil {
 			return logs, nil
 		}
 
-		logdog.Infof("checkout code as error %v, retry times: %v", err, i)
-		time.Sleep(time.Second * 1)
+		logdog.Infof("failed to checkout, will retry after %d seconds, due to error: %v", backoff, err)
+		time.Sleep(backoff)
+		backoff *= 2
```

Comment:
What about `backoff += 2` ?
We will wait for almost 512(2 to the 9th power) seconds in the worst case, and i think it is a little long.

Reply:
It's useless to use `backoff += 2` to retry, in most condition, a service needs more time to recovery from downtime.

Reply:
retry is mainly for unstable environment, not for service recovery.

Reply:
Just like `pending` in k8s, we give the program a status or a goal, it keeps on trying to make it

Reply:
The `pending` pipelines will hold the resources, then other pipelines will be blocked because lack of resource. 
On the other side, if the pipeline will fail in the end, it is not very good to give the feedback after a long wait.

So, I think retry within a short period is reasonable.